File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/emulations/rust/i32_to_f64.rs
#![allow(non_snake_case, unused_parens, unused_imports, unused_variables, unused_comparisons, clippy::all)]
use crate::helpers::*;


pub fn i32_to_f64_rm1(v_arg: u32) -> u64 {
    let v_i: bool = (v_arg == 0x0u32);
    let v__n7: bool = (v_i ^ true);
    let v__m8: u64 = (0u64.wrapping_sub((v__n7) as u64));
    let v__k1: u32 = sf_abs_u32(v_arg);
    let v__k2: u32 = sf_ctlz_u32(v__k1);
    let v_i3: u32 = v__k2;
    let v_i12: u64 = ((v__k1) as u64);
    let v_i4: u32 = v_i3.wrapping_add(0x15u32);
    let v_i8: u32 = 0x41du32.wrapping_sub(v_i3);
    let v_i9: u64 = ((v_i8) as u64);
    let v__sh5: u64 = v_i9.wrapping_shl((0x34u64) as u32);
    let v_i13: u64 = ((v_i4) as u64);
    let v__sh6: u64 = v_i12.wrapping_shl((v_i13) as u32);
    let v_i14: u64 = v__sh6;
    let v__sh3: u32 = v_arg.wrapping_shr((0x1fu32) as u32);
    let v_i6: u64 = ((v__sh3) as u64);
    let v__sh4: u64 = v_i6.wrapping_shl((0x3fu64) as u32);
    let v_i11: u64 = (v__sh5 | v__sh4);
    let v_i15: u64 = v_i11.wrapping_add(v_i14);
    let v__a9: u64 = (v_i15 & v__m8);
    v__a9
}

File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/emulations/python/i32_to_f64.py
from helpers import (sf_udiv, sf_sgn, sf_ctlz, sf_abs,
                     sf_usubsat, sf_fshl)


def i32_to_f64_rm1(v_arg):
    v_i = (1 if v_arg == 0x0 else 0)
    v__n7 = (v_i ^ 0x1)
    v__m8 = (sf_sgn(v__n7, 0x1) & 0xffffffffffffffff)
    v__k1 = sf_abs(v_arg, 0x20)
    v__k2 = sf_ctlz(v__k1, 0x20)
    v_i3 = v__k2
    v_i12 = (v__k1)
    v_i4 = ((v_i3 + 0x15) & 0xffffffff)
    v_i8 = ((0x41d - v_i3) & 0xffffffff)
    v_i9 = (v_i8)
    v__sh5 = ((v_i9 << (0x34 & 0x3f)) & 0xffffffffffffffff)
    v_i13 = (v_i4)
    v__sh6 = ((v_i12 << (v_i13 & 0x3f)) & 0xffffffffffffffff)
    v_i14 = v__sh6
    v__sh3 = (v_arg >> (0x1f & 0x1f))
    v_i6 = (v__sh3)
    v__sh4 = ((v_i6 << (0x3f & 0x3f)) & 0xffffffffffffffff)
    v_i11 = (v__sh5 | v__sh4)
    v_i15 = ((v_i11 + v_i14) & 0xffffffffffffffff)
    v__a9 = (v_i15 & v__m8)
    return v__a9
